flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeGeneralization.java

```diff
@@ -0,0 +1,631 @@
+		if (typeRoot == TIMESTAMP_WITH_TIME_ZONE ||
+				resultTypeRoot == TIMESTAMP_WITH_TIME_ZONE) {
+			return createTimestampType(TIMESTAMP_WITH_TIME_ZONE, precision);
+		} else if (typeRoot == TIMESTAMP_WITH_LOCAL_TIME_ZONE ||
+				resultTypeRoot == TIMESTAMP_WITH_LOCAL_TIME_ZONE) {
+			return createTimestampType(TIMESTAMP_WITH_LOCAL_TIME_ZONE, precision);
+		}
+		return createTimestampType(TIMESTAMP_WITHOUT_TIME_ZONE, precision);
+	}
+
+	private static LogicalType createCommonDayTimeIntervalType(DayTimeIntervalType resultType, DayTimeIntervalType type) {
+		final int maxDayPrecision = Math.max(resultType.getDayPrecision(), type.getDayPrecision());
+		final int maxFractionalPrecision = Math.max(resultType.getFractionalPrecision(), type.getFractionalPrecision());
+		return new DayTimeIntervalType(
+			combineIntervalResolutions(
+				DayTimeResolution.values(),
+				DAY_TIME_RES_TO_BOUNDARIES,
+				DAY_TIME_BOUNDARIES_TO_RES,
+				resultType.getResolution(),
+				type.getResolution()),
+			maxDayPrecision,
+			maxFractionalPrecision);
+	}
+
+	private static LogicalType createCommonYearMonthIntervalType(YearMonthIntervalType resultType, YearMonthIntervalType type) {
+		final int maxYearPrecision = Math.max(resultType.getYearPrecision(), type.getYearPrecision());
+		return new YearMonthIntervalType(
+			combineIntervalResolutions(
+				YearMonthResolution.values(),
+				YEAR_MONTH_RES_TO_BOUNDARIES,
+				YEAR_MONTH_BOUNDARIES_TO_RES,
+				resultType.getResolution(),
+				type.getResolution()),
+			maxYearPrecision);
+	}
+
+	private static LogicalType createTimestampType(LogicalTypeRoot typeRoot, int precision) {
+		switch (typeRoot) {
+			case TIMESTAMP_WITHOUT_TIME_ZONE:
+				return new TimestampType(precision);
+			case TIMESTAMP_WITH_TIME_ZONE:
+				return new ZonedTimestampType(precision);
+			case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+				return new LocalZonedTimestampType(precision);
+			default:
+				throw new IllegalArgumentException();
+		}
+	}
+
+	private static LogicalType createStringType(LogicalTypeRoot typeRoot, int length) {
+		switch (typeRoot) {
+			case CHAR:
+				return new CharType(length);
+			case VARCHAR:
+				return new VarCharType(length);
+			case BINARY:
+				return new BinaryType(length);
+			case VARBINARY:
+				return new VarBinaryType(length);
+			default:
+				throw new IllegalArgumentException();
+		}
+	}
+
+	private static <T extends Enum<T>> T combineIntervalResolutions(
+			T[] res,
+			Map<T, List<T>> resToBoundaries,
+			Map<List<T>, T> boundariesToRes,
+			T left,
+			T right) {
+		final List<T> leftBoundaries = resToBoundaries.get(left);
+		final T leftStart = leftBoundaries.get(0);
+		final T leftEnd = leftBoundaries.get(leftBoundaries.size() - 1);
+
+		final List<T> rightBoundaries = resToBoundaries.get(right);
+		final T rightStart = rightBoundaries.get(0);
+		final T rightEnd = rightBoundaries.get(rightBoundaries.size() - 1);
+
+		final T combinedStart = res[Math.min(leftStart.ordinal(), rightStart.ordinal())];
```

Comment:
If we want to depend on the ordinal of `Resolution`s let's at least add a comment next to those classes that the order has to be maintained.

To be honest though, I would feel safer if those enums implemented `Comparable`.